source/computer/commands.rst

```diff
@@ -63,54 +62,505 @@ Linux/macOS 下有成百上千个命令，每个命令都有众多选项。这
+    $ find ~/src -name "*.c"
+
+    # 查找 ~/src 目录及其子目录下以 .c 结尾的文件
+    $ find ~/src -type f -name "*.c"
+
+使用 ``-delete`` 选项可以直接删除查找的文件或目录::
+
+    # 查找 ~/src 目录及其子目录下以 .pyc 结尾的文件，并删除
+    $ find ~/src -type f -name "*.pyc" -delete
+
+``find`` 的 ``-exec`` 选项可以调用其他系统命令直接对查找的结果进行处理::
+
+    # 查找 ~/src 目录及其子目录下以 .c 结尾的文件，并执行 grep 命令找出这些文件中含 seismo-learn 的行
+    $ find ~/src -type f -name "*.c" -exec grep seismo-learn {} +
+
+这个例子中，``{}`` 与 ``-exec`` 选项结合，表示查找到的文件。
+我们还可以用 ``-ok`` 来代替 ``-exec`` 选项，二者的区别是 ``-ok`` 选项在执行后面的命令前会给出提示，
+输入 :kbd:`y` 才会执行，输入 :kbd:`n` 则不执行。
+
 gawk
 ----
 
+``awk`` 命令的命名起源于其三位作者的姓氏首字母。该命令可以选择标准输入、其他命令的输出或文件中
+的特定字段并进行操作。它依次扫描每一行，并读取里面的每一个字段。可以参考
+`awk 入门教程 <https://www.ruanyifeng.com/blog/2018/11/awk.html>`__ 学习其更多用法。
+
+``gawk`` 是 GNU 版本的 ``awk`` 命令。通常建议直接使用 ``gawk`` 而非 ``awk``，
+尽管在 Linux 系统下，``awk`` 命令一般是指向 ``gawk`` 命令的软链接。
+
+::
+
+    # 将一个文件的每一行输出到标准输出。 $0 表示当前行
+    $ gawk '{print $0}' file
+
+    # 将一个文件的每一行的第 1 个和第 3 个字段输出到标准输出（字段的默认分隔符是空格和制表符）
+    $ gawk '{print $1,$3}' file
+
+    # 同上，但跳过第 1-2 行，从第 3 行开始。FNR 为当前行的行数
+    $ gawk 'FNR>2 {print $1,$3}' file
+
+    # 将 /etc/passwd 每一行的第 1 个和第 3 个字段输出到标准输出，并设置字段分隔符为冒号
+    $ gawk -F ':' '{print $1,$3}' /etc/passwd
+
+    # 同上，并输出每一行的行号
+    $ gawk -F ':' '{print NR") " $1,$3}' /etc/passwd
+
 grep
 ----
 
+``grep`` 命令的命名来自 **g**\ lobally search a **re**\ gular expression and **p**\rint
+（以正则表达式进行全局搜索并输出）。它是一个强大的文本搜索工具，可以搜索文件中与指定模式匹配的行并输出::
+
+    # 搜索并输出 /etc/passwd 中含有 root 的行
+    $ grep root /etc/passwd
+
+    # 同上，并显示行号
+    $ grep -n root /etc/passwd
+
+    # 搜索并输出 /etc/passwd 中不含有 root 的行
+    $ grep -v root /etc/passwd
+
 head
 ----
 
+``head`` 命令用于输出文件开头部分的内容::
+
+    # 输出 /etc/passwd 的前 10 行
+    $ head /etc/passwd
+
+    # 输出 /etc/passwd 的前 5 行
+    $ head -n 5 /etc/passwd
+
 less
 ----
 
+``less`` 命令可以用来浏览文件内容，比 ``more`` 命令功能更强大。因此，一个常用的笑话记忆方式是
+“less is more”。该命令允许用户向前或向后浏览文件：按 :kbd:`PageUp` 键向上翻页，
+用 :kbd:`PageDown` 键向下翻页，按 :kbd:`Enter` 键或向下方向键则向下移动，
+用向上方向键则向上移动。按 :kbd:`q` 键退出浏览。
+
+::
+
+    # 浏览 /etc/passwd
+    $ less /etc/passwd
+
+    # 同上，并显示行号
+    $ less -N /etc/passwd
+
 locate
 ------
 
+``locate`` 命令可以用于查找目录和文件。该命令比 `find`_ 命令快得多，原因在于它不搜索具体目录，
+而是直接搜索含有本地所有文件信息的数据库::
+
+    # 搜索名称含 passwd 的目录或文件
+    $ locate passwd
+
+    # 搜索名称含 /etc/sh 的目录或文件（包括了 /etc 目录下所有以 sh 开头的目录或文件）
+    $ locate /etc/sh
+
+    # 搜索名称含 ~/Des 的目录或文件（包括了家目录下，所有以 Des 开头的目录或文件）
+    $ locate ~/Des
+
+    # 同上，但忽略大小写
+    $ locate -i ~/des
+
+该命令所需的数据库是系统自动创建的，每天自动更新。因此，``locate`` 命令查不到最新变动过的文件。
+可以执行 ``updatedb`` 命令手动更新数据库，但是 ``updatedb`` 命令的执行过程较长::
+
+    # Linux
+    $ sudo updatedb
+
+    # macOS 系统可以使用以下命令
+    $ sudo /usr/libexec/locate.updatedb
+
 rsync
 -----
 
+``rsync`` 命令的命名来自 **r**\ emote **sync**\ hronization（远程同步）。该命名是一个数据同步工具，
+可以在两个本地目录之间，或本地计算机与远程计算机之间同步文件。与其他文件传输工具（如 FTP 或 `scp`_\ ）不同，
+``rsync`` 命名最大特点是会检查发送方和接收方已有的文件，仅传输有变动的部分（默认规则是文件大小或
+修改时间有变动），因此速度较快，且常用于文件备份。可以参考
+`rsync 用法教程 <https://www.ruanyifeng.com/blog/2020/08/rsync.html>`__
+进一步学习其用法。
+
```

Comment:
Some thoughts about `rsync`:

1. The most common and useful command is `rsync -av --delete`. I think we should only mention this one. Why do we need `rsync -av` without `--delete`?
2. I understand the differences between `~/Downloads/source` and `~/Downloads/source/`, but it's unclear to me which one is more commonly used.
3. What I expect are the 4 commands:

```
$ rsync -av --delete ~/Downloads/source ~/workspace/destination
$ rsync -av --delete ~/Downloads/source/ ~/workspace/destination
$ rsync -av --delete ~/Downloads/source seismo-learn@192.168.1.100:~/destination
$ rsync -av --delete seismo-learn@192.168.1.100:~/source ~/workspace/destination
```

Reply:
> Some thoughts about `rsync`:
> 
> 1. The most common and useful command is `rsync -av --delete`. I think we should only mention this one. Why do we need `rsync -av` without `--delete`?

Yes to me.

> 2. I understand the differences between `~/Downloads/source` and `~/Downloads/source/`, but it's unclear to me which one is more commonly used.

The first one is used in file backup and transfer.

The second one is used to make a mirror.

> 3. What I expect are the 4 commands:
> 
> ```
> $ rsync -av --delete ~/Downloads/source ~/workspace/destination
> $ rsync -av --delete ~/Downloads/source/ ~/workspace/destination
> $ rsync -av --delete ~/Downloads/source seismo-learn@192.168.1.100:~/destination
> $ rsync -av --delete seismo-learn@192.168.1.100:~/source ~/workspace/destination
> ```

Yes to me, but shall we add an example with `-anv`?

Reply:
check 88f3eb2.